### ISSUE TITLE
Improve Tic-Tac-Toe DQN so it stops “dreaming” about illegal moves during learning and learns under pressure from a blocking/winning opponent. Also extend training & tweak a few hyper-params. Evaluate with the existing scripts.

### ISSUE THREAD

**Issue by dvroegop** [dvroegop/tictactoe_python#13]
Improve Tic-Tac-Toe DQN so it stops “dreaming” about illegal moves during learning and learns under pressure from a blocking/winning opponent. Also extend training & tweak a few hyper-params. Evaluate with the existing scripts.

Changes by file
1) tictactoe_package/dqn_agent.py

Goal: mask illegal actions in the learning target (we already mask at inference), and optionally support Double-DQN later. Right now learn() bootstraps with max_a' Q_target(s',a') over all 9 actions, including illegal ones. Fix that.

A. Extend replay tuples to carry the next-state legal mask

Add an optional mask_next to the replay buffer.

# Replay memory item becomes: (s, a, r, s_next, done, mask_next)
def remember(self, s, a, r, s_next, done, mask_next=None):
    self.buffer.append((s, a, r, s_next, done, mask_next))

B. Use mask_next inside learn() before the max over a′

Inside learn():

batch = random.sample(self.buffer, self.cfg.batch_size)
s, a, r, s_next, done, mask_next = zip(*batch)

s      = torch.stack(s).to(self.cfg.device)
a      = torch.tensor(a, dtype=torch.long, device=self.cfg.device)
r      = torch.tensor(r, dtype=torch.float32, device=self.cfg.device)
s_next = torch.stack(s_next).to(self.cfg.device)
done   = torch.tensor(done, dtype=torch.float32, device=self.cfg.device)

with torch.no_grad():
    q_next_all = self.target(s_next)  # (B,9)
    if mask_next[0] is not None:
        mask_next = torch.stack(mask_next).to(self.cfg.device)  # (B,9) 0/1
        q_next_all = q_next_all.masked_fill(mask_next < 0.5, -1e9)
    q_next = q_next_all.max(1).values
    target = r + self.cfg.gamma * q_next * (1.0 - done)

Note: You already mask illegal actions at selection time in select_action / pick_move; leave that as is. This change fixes the bootstrapping path.

(Optional) Double-DQN target (keep behind a flag)

If you want, add a config flag and compute a* with the online net, then gather from target; still honor mask_next.

2) train_dqn.py

Goal: supply mask_next to replay, train longer with slower ε-decay, and periodically train against a smart opponent.

A. Import legal_mask and pass it to remember

After every transition, compute the next state legal mask:

from tictactoe_package.dqn_agent import DQNAgent, DQNConfig, encode_board, legal_mask

In both the non-terminal and terminal branches, when calling agent.remember(...), add mask_next=legal_mask(env.board) (note: that’s the legal mask of s_next as seen by env.current_player after you switch player when applicable).

Non-terminal path (you already do env.switch_player() before building s_next):

env.switch_player()
r = step_penalty
s_next = encode_board(env.board, env.current_player)
mask_next = legal_mask(env.board)
agent.remember(s, a, r, s_next, False, mask_next)

Terminal path: record terminal snapshot and mask (it won’t be used for bootstrapping because done=True, but keep it consistent):

s_next = encode_board(env.board, env.current_player)
mask_next = legal_mask(env.board)
agent.remember(s, a, r, s_next, True, mask_next)

Your “losing move” extra write for the previous player (good idea—keep it) should also pass a mask (same terminal board):

prev_s_next = encode_board(env.board, prev_mover)
prev_mask_next = legal_mask(env.board)
agent.remember(prev_s, prev_a, prev_r, prev_s_next, True, prev_mask_next)

B. Add a “smart opponent curriculum” for part of the episodes

You already have smart opponent logic in evaluate_dqn_smart.py (smart_opponent_move, check_winning_move). Reuse it during training for, say, 25–50% of episodes to increase pressure.

Easiest: copy those two functions into train_dqn.py (or import) and add:

use_smart = (ep % 4 == 0)  # every 4th episode

When it’s the opponent’s turn (i.e., when the DQN did not just move), choose:

if use_smart and env.current_player != mover_just_used_by_dqn:
    position = smart_opponent_move(env.board, env.current_player)
else:
    # keep current behavior (DQN controls both sides), i.e., select_action(..., explore=True)

Keep rewards, remember, masks, and learning calls exactly as for DQN-vs-DQN steps so the replay distribution includes tough counter-play.

C. Training length & small hyper-param nudges

Increase episodes from 8000 → 30000 (or make it a CLI arg).

In DQNConfig:

epsilon_decay_steps: 20000 (from 5000) to avoid collapsing exploration too soon.

target_sync_every: 200 (from 500) for crisper targets.

(Optional) batch_size: 64 is fine; 128 is okay if stable.

No other reward shaping changes needed (you already use a small step penalty and ±1 outcomes).

3) (No code change) controller.py

Nothing required here—the controller already falls back to a random legal move if pick_move returns something illegal. Once learning is fixed, that path should almost never trigger.

4) Evaluation

Run both evaluators; they already exist:

Self/Random: evaluate_dqn.py
Targets printed at bottom: high draw in DQN vs DQN, high win/draw vs Random.

Smart Opponent: evaluate_dqn_smart.py
Targets printed at bottom: many draws (≥40%) and rare losses (<30%).

(If results are off, enable cfg.verbose=True to inspect the action scores; the “Top candidates” print is already wired.)

Acceptance criteria

Illegal-move masking in targets

Confirm remember stores 6-tuples and learn() masks q_next_all using mask_next, then takes max.

Curriculum enabled

Every ~4th episode uses the smart opponent for the side not controlled by the DQN.

Training still records transitions (including the “losing move” of the previous player) exactly as before.

Hyper-params updated

epsilon_decay_steps=20000, target_sync_every=200, episodes=30000.

Observed performance (100-game blocks)

DQN vs DQN: draw rate ≥ 70%.

DQN vs Random: win+draw ≥ 90%.

DQN vs Smart: draws ≥ 40%, DQN losses ≤ 30%.

Notes & pitfalls

When you compute mask_next, be careful which player’s turn it is. Use the environment’s current board immediately after you set s_next, which already reflects the next mover in your loop.

Keep terminals flagged done=True; the mask won’t affect targets on those samples, but consistency simplifies batching.

If loss spikes or oscillates, clip gradients to 1.0 and/or reduce LR to 5e-4.
